frontend/apps/crates/components/src/module/_groups/cards/edit/sidebar/step_1/state.rs

```diff
@@ -18,38 +18,48 @@ use crate::{
 };
 use dominator::clone;
 use futures_signals::signal::Mutable;
+use once_cell::sync::OnceCell;
 use shared::domain::jig::module::body::{Image, _groups::cards::Mode};
 use std::rc::Rc;
 
 pub struct Step1<RawData: RawDataExt, E: ExtraExt> {
     pub base: Rc<CardsBase<RawData, E>>,
-    pub widget: Rc<Widget>,
+    pub widget: OnceCell<Widget>,
     pub tab_index: Mutable<Option<usize>>,
 }
 
 impl<RawData: RawDataExt, E: ExtraExt> Step1<RawData, E> {
     pub fn new(base: Rc<CardsBase<RawData, E>>, tab_index: Mutable<Option<usize>>) -> Rc<Self> {
+        let state = Rc::new(Self {
+            base: base.clone(),
+            widget: OnceCell::default(),
+            tab_index,
+        });
+
+        // Widgets require a reference to the top-level state so that they can have access to any
+        // fields they might require in callbacks.
```

Comment:
This comment relates specifically to [this change](https://github.com/ji-devs/ji-cloud/pull/2072/files#diff-a6de13114d3690edd188660e6ee5f53a9003eb914845d9918f2a17d978dd88a0R132-R134)